### ISSUE TITLE
No subcomponents found if there is component registered multiple times

### ISSUE THREAD

**Issue by lulco** [efabrica-team/phpstan-latte#377]
If component of some type and same name is registered more than once, we somehow lost its subcomponents
```
if ($someCondition) {
    $presenter->addComponent(new SomeControl(), 'someControl');
} else {
    $presenter->addComponent(new SomeControl(), 'someControl');
}
```

Failing test case https://github.com/efabrica-team/phpstan-latte/pull/376

I think there is some problem in ItemCombinator or ComponentFinder (more probably ItemCombinator).

@MartinMystikJonas any idea what it can be?

maybe overwriting component when the name is the same:
```
foreach ($itemArray as $item) {
    $itemName = $item->getName();
    $merge[$itemName] = $item;
}
```
?

**Comment by MartinMystikJonas** [comment]
Replacing items with same name in emrge is excpected behaviour. I think it should not affect subcomponents.

My guess is there would be some bug in ComponentFinder

**Comment by lulco** [comment]
yes that's the second place where we have to look at

**Comment by MartinMystikJonas** [comment]
I think I know where the problem is. In ComponentFinder we assign components to $componentsWithTypes before merge/union is made. And we set sub-components for these components stored in $componentsWithTypes only. But union can create new Component object with union type if same name is present multiple times. This new Component object is not in $componentsWithTypes - old versions with non-union types are there. It is problem even when type is the same because we do not hceck if crating new Component object is necessary

**Comment by MartinMystikJonas** [comment]
I think logic of subcomponents in ComponentFinder  needs to work differently. Current implementation would fail with multi-add component qith same name and also with adding different types or union type.

I think we need to first create lists of declared/assigned components, and then go throught this whole lists to find sub-components where we use some kind of isSupertypeOf logic.

**Comment by MartinMystikJonas** [comment]
Managed to fixed that. I will commit it soon - I have to fix some PHPStan errors

**Comment by MartinMystikJonas** [comment]
Fixed in #376

**Comment by lulco** [comment]
I'm going to check it
